README.md
# mondrian: A modular, extensible platform for scWGS informatics

For a detailed guide see [INSTALL](docs/source/index.md)


[Changelog](CHANGELOG.md)

## What is it?

Mondrian is a proposed suite of workflows for analysing the single cell data generated by single cell sequencing platforms. It’ll provide single cell level workflows for aligning the fastq files to the human reference genome and to infer copy number at a single cell level. Mondrian will also provide pseudo-bulk level workflows to look for point mutations, indels and structural variations. Mondrian will also provide genotyping workflows to overlay the pseudo bulk results on a single cell level


## Where to get it
The source code is currently hosted on GitHub at:
https://github.com/shahcompbio/mondrian

docker containers are available at
https://hub.docker.com/orgs/mondrian

conda packages are available at
https://anaconda.org/mondrian

## License
[GPL v3.0](LICENSE)

## Documentation
The official documentation is hosted on https://mondrian.readthedocs.io/en/latest/

## Getting Help

For usage questions, the best place to go to is [Google Groups](https://groups.google.com/forum/#!forum/mondrianscdna).

## Discussion and Development
Most development discussions take place on github in this repo.
 and a [Gitter channel](https://gitter.im/mondrianscdna) is available for quick development related questions.

## Contributing to mondrian

The code is hosted on [GitHub](https://www.github.com/shahcompbio/mondrian), so you will need to use [Git](https://git-scm.com/) to clone the project and make changes to the codebase. Once you have obtained a copy of the code, you should create a development environment that is separate from your existing Python environment so that you can make and test changes without compromising your own work environment.

Before submitting your changes for review, make sure to check that your changes do not break any tests.

Once your changes are ready to be submitted, make sure to push your changes to GitHub before creating a pull request. We will review your changes, and you will most likely be asked to make additional changes before it is finally ready to merge. However, once it's ready, we will merge it, and you will have successfully contributed to the codebase!


As contributors and maintainers to this project, you are expected to abide by mondrian' code of conduct. More information can be found at: [Contributor Code of Conduct](https://github.com/shahcompbio/mondrian/blob/master/.github/CODE_OF_CONDUCT.md).

Feel free to ask questions on the mailing list or on Gitter.

As contributors and maintainers to this project, you are expected to abide by mondrian's code of conduct. More information can be found at: Contributor Code of Conduct


CHANGELOG.md
# Change Log

### v0.7.0

#### Changes:
1. rename to mondrian

### v0.6.20

### Changes:
1. conda packages are now built on top on py 3.8

### Bug:
1. minor heatmap axis labeling bug (assumes chrom1 is starting point)


### v0.6.19

#### Changes:
1. produces a MT bam file
2. setup doc also includes information to switch pipeline to production datasets


### v0.6.18

#### Bugs:
1. missing data error in species classifier
2. build: pip cannot install from commits in a PR coming from a fork

#### Changes:
1. conda: annotation pipeline requires jinja2 dependency
2. conda: align requires ete3 and R
2. added a quickstart guide based on conda

#### v0.6.17

#### Bugs:
1. missing fastqscreen training data for mouse in config
2. data type bug in fastqscreen summary

### v0.6.16

#### Changes:
1. new build process (AWS codebuild, on PR)


### v0.6.15

#### Bug:
1. there were no counts for cells with no data in fastqscreen summary

### v0.6.14

#### Changes:
1. removed joblib files from repo
2. added conda recipes

### v0.6.13

#### Changes:
1. updated remixt


### v0.6.12

#### Bug:
1. .tmp in input.yaml file path in  metadata yaml file

#### Changes:
1. test data set for count haps
2. updated infer haps testing
3. removed hardcoded genomes in HTML report
4. infer and count haps use proper chromosomes

### v0.6.11

#### Changes:
1. pypeliner version update to v0.5.19



### v0.6.10

#### Bug:
1. missing cell id in count haps

### v0.6.9

#### Bug:
1. fixed snv annotation issues

#### Changes:
1. biowrappers upgraded to v0.2.8

### v0.6.8

#### Bug:
1. incorrect rebase on is_contaminated flag code
2. fixed issues with tag name in docker build

#### Changes:
1. updated QC generation code, added classifier
2. updated docs: read the docs compatible

### v0.6.7

#### Bug:
1. issue with dtypes for lumpy

#### Changes:
1. updated biowrappers to v0.2.7

### v0.6.6

#### Bug:
1. missing yaml extension for count haps input


### v0.6.5

#### Bug:
1. issues with missing yaml files

#### Changes:
1. add yaml validators
2. added docs


### v0.6.4:

#### Bug:
1. Issue with fastqscreen counts
2. remove corrupt tree files from metadata
3. dtype for chr is str not int in lumpy

### v0.6.3:

#### Bug:
1. issue with yaml missing in extensions for snv file

### v0.6.2:

#### Bug:
1. collect metrics introduced nan into int columns

### v0.6.1

### Changes:
1. per sample genotyping results

### v0.6.0

### Changes:
1. jenkins refactor
2. csvutils refactor

### V0.5.21

#### Changes:
1. upgraded remixt to v0.5.11

### v0.5.20

#### Changes:
1. upgraded destruct and remixt to latest pypeliner.

### v0.5.19

####Changes:
1. upgraded remixt and cell cycle classifier

### v0.5.18

#### Changes:
1. upgraded pypeliner to v0.5.18

### v0.5.17

#### Changes:
1. updated the calculation for is_contaminated
2. moved contaminated flagging to annotation
3. pypeliner updated to v0.5.17
4. removed org from scp docker container names in config

### v0.5.16

####Changes:
1. updated pypeliner to v0.5.16
2. conda build fixed


### v0.5.15

#### bug:
1. metadata regions are none sometimes

### v0.5.14

#### bug:
1. hdf to csv was only writing header for first output file


### v0.5.13

#### bug:
1. splitting heatmap into 1000 cells per page doesnt account for cases where the next page only has 1 cell which cant be plotted.

### v0.5.12

#### bug:
1. destruct read indexing was broken

#### Changes:
1. flags to selectively run destruct or lumpy

### v0.5.11:

#### Changes:
1. tempdir is now separated by subcommands
2. file names updated for breakpoint calling


### v0.5.10

#### Bug:
1. conversion to csv from hdf was overwriting not appending

#### Changes:
1. destruct version updated
2. docker container is built from tag, not from commit id

### v0.5.9

#### Bug:
1. excessive memory usage in h5 to csv conversion
2. fixed config issue for snpeff

### v0.5.8

#### Changes:
1. updated biowrappers to v0.2.4
2. updated pypeliner version to v0.5.15
3. optimized QC VM size for cost savings
4. snpeff now uses data from the refdata dir instead of downloading
5. added test datasets


### v0.5.7

#### Changes:
1. updated biowrappers version to v0.2.2
2. updated destruct and remixt container versions
3. added sv genotyping (experimental)

#### bugs:
1. missing last line in lumpy parsed output
2. issues with fastqscreen tags 


### v0.5.6:

#### bugs:
1. fixed an issue with double headers in snv calling output
2. issues in plotting due to gc data dtypes set to string

#### Changes:
1. destruct container does not need single cell pipeline
2. remixt container does not need single cell pipeline 
3. some updates to documentation
4. updated to biowrappers v0.2.1
5. split infer and count haps

### v0.5.5:

#### changes:
1. updated pypeliner to v0.5.13
2. updated documentation

### v0.5.4:

#### bugs:
1. mismatching csv type error in empty fastq files

#### changes:
1. refactor errors with empty fastq screen files when fastq is empty


### v0.5.3:

#### bugs:
1. fixed base64 encode issue for images in qc reports
2. issue deleting non-existent bam key
3. missing .gz.csv.yaml extensions for lumpy files
4. fixed container for lumpy
5. pd.concat on empty set of dataframes fixed
6. plotting bugs in hmmcopy
7. pandas loading issues related to dtypes specified but not names
8. use new hmmcopy script, better params format

### v0.5.2:

#### Changes:
1. alignment for all lanes is run in a single job
2. optimized destruct fastq reindexing

#### bugs:
1. removed redundant replace ? by 0 in plotting
2. annotation works when sample_info is none
3. fixed paths for annotation low mad yaml in tests
4. updated dtypes in integrity tests

#### bugs:
1. issues with NA handling in csvutils
2. all median cols are float now

### v0.5.1:

#### bugs:
1. issues with NA handling in csvutils
2. all median cols are float now

### v0.5.0: 
CLI refactor

#### added:
1. germline calling mode

#### Changes:
1. refactored the CLI
2. removed: 
    a) QC and

tests/jenkins/hmmcopy/hmmcopy.sh
#!/bin/bash
set -e
set -o pipefail

TAG=`git describe --tags $(git rev-list --tags --max-count=1)`
TAG="${TAG}.beta"


mkdir -p HMMCOPY/ref_test_data

docker run -v $PWD:$PWD -w $PWD $3/azurecli:v0.0.1 \
  az storage blob download-batch -s hmmcopy -d HMMCOPY/ref_test_data --account-name $1 --account-key $2

docker run -w $PWD -v $PWD:$PWD -v /refdata:/refdata -v /var/run/docker.sock:/var/run/docker.sock \
  -v /usr/bin/docker:/usr/bin/docker --rm \
  $3/single_cell_pipeline:$TAG \
  single_cell hmmcopy --input_yaml tests/jenkins/hmmcopy/inputs.yaml \
  --library_id A97318A --maxjobs 4 --nocleanup --sentinel_only  \
  --context_config tests/jenkins/context_config.yaml \
  --submit local --loglevel DEBUG \
  --config_override '{"hmmcopy": {"chromosomes": ["6", "8", "17"]}}' --tmpdir HMMCOPY/temp \
  --pipelinedir HMMCOPY/pipeline --submit local --out_dir HMMCOPY/output

docker run -w $PWD -v $PWD:$PWD -v /refdata:/refdata -v /var/run/docker.sock:/var/run/docker.sock \
  -v /usr/bin/docker:/usr/bin/docker --rm \
  $3/single_cell_pipeline:$TAG \
  python tests/jenkins/hmmcopy/test_hmmcopy.py HMMCOPY/output A97318A  HMMCOPY/ref_test_data/refdata


tests/jenkins/annotation/annotation.sh
#!/bin/bash
set -e
set -o pipefail

TAG=`git describe --tags $(git rev-list --tags --max-count=1)`
TAG="${TAG}.beta"

mkdir -p ANNOTATION/ref_test_data

docker run -v $PWD:$PWD -w $PWD $3/azurecli:v0.0.1 \
  az storage blob download-batch -s annotation -d ANNOTATION/ref_test_data --account-name $1 --account-key $2


docker run -w $PWD -v $PWD:$PWD -v /refdata:/refdata -v /var/run/docker.sock:/var/run/docker.sock \
  -v /usr/bin/docker:/usr/bin/docker --rm \
  $3/single_cell_pipeline:$TAG \
  single_cell annotation --input_yaml tests/jenkins/annotation/inputs.yaml \
  --library_id A97318A --maxjobs 4 --nocleanup --sentinel_only  \
  --context_config tests/jenkins/context_config.yaml \
  --submit local --loglevel DEBUG \
  --tmpdir ANNOTATION/temp \
  --pipelinedir ANNOTATION/pipeline --submit local --out_dir ANNOTATION/output \
  --config_override '{"annotation": {"chromosomes": ["6", "8", "17"]}}' --no_corrupt_tree

docker run -w $PWD -v $PWD:$PWD -v /refdata:/refdata -v /var/run/docker.sock:/var/run/docker.sock \
  -v /usr/bin/docker:/usr/bin/docker --rm \
  $3/single_cell_pipeline:$TAG \
  python tests/jenkins/annotation/test_annotation.py ANNOTATION/output A97318A  ANNOTATION/ref_test_data/refdata


tests/jenkins/count_haps/count_haps.sh
#!/bin/bash
set -e
set -o pipefail

TAG=`git describe --tags $(git rev-list --tags --max-count=1)`
TAG="${TAG}.beta"

mkdir -p COUNT_HAPS/ref_test_data

docker run -v $PWD:$PWD -w $PWD singlecellpipeline/azurecli:v0.0.1 \
  az storage blob download-batch -s count-haps  -d COUNT_HAPS/ref_test_data --account-name $1 --account-key $2


docker run -w $PWD -v $PWD:$PWD -v /refdata:/refdata -v /var/run/docker.sock:/var/run/docker.sock \
  -v /usr/bin/docker:/usr/bin/docker --rm \
  $3/single_cell_pipeline:$TAG \
  single_cell count_haps --input_yaml tests/jenkins/count_haps/inputs.yaml \
  --maxjobs 4 --nocleanup --sentinel_only  \
  --context_config tests/jenkins/context_config.yaml \
  --submit local --loglevel DEBUG \
  --tmpdir COUNT_HAPS/temp \
  --config_override '{"count_haps":{"chromosomes":["15"], "extract_seqdata": {"genome_fai_template": "/refdata/human/infer_haps/GRCh37-lite.fa.fai", "genome_fasta_template": "/refdata/human/infer_haps/GRCh37-lite.fa"}, "ref_data_dir": "/refdata/human/infer_haps/"}, "version": '\"$TAG\"'}' \
  --pipelinedir COUNT_HAPS/pipeline --submit local --out_dir COUNT_HAPS/output

docker run -w $PWD -v $PWD:$PWD -v /refdata:/refdata -v /var/run/docker.sock:/var/run/docker.sock \
  -v /usr/bin/docker:/usr/bin/docker --rm \
  $3/single_cell_pipeline:$TAG \
  python tests/jenkins/count_haps/test_count_haps.py COUNT_HAPS/output COUNT_HAPS/ref_test_data/refdata


single_cell/config/config_reference.py
import os

from single_cell.utils import helpers


def containers(version=None):
    if not version:
        version = helpers.get_version()
        version = 'v'+version

    docker_images = {
        'bwa': 'bwa:v0.0.2',
        'samtools': 'samtools:v0.0.3',
        'python_base': 'python_base:v0.1.0',
        'picard': 'picard:v0.0.3',
        'single_cell_pipeline': 'single_cell_pipeline:{}'.format(version),
        'gatk': 'gatk:v0.0.1',
        'fastqc': 'fastqc:v0.0.2',
        'hmmcopy': 'hmmcopy:v0.0.5',
        'aneufinder': 'aneufinder:v0.0.1',
        'strelka': 'strelka:v0.0.3',
        'mutationseq': 'mutationseq:v0.0.3',
        'vcftools': 'vcftools:v0.0.2',
        'snpeff': 'vcftools:v0.0.2',
        'titan': 'titan:v0.0.1',
        'remixt': 'remixt:v0.0.3',
        'destruct': 'destruct:v0.0.4',
        'trimgalore': 'trimgalore:v0.0.2',
        'lumpy': 'lumpy:v0.0.3',
        'cell_cycle_classifier': 'cell_cycle_classifier:v0.0.1',
        'biobloom': 'biobloom:v0.0.2',
        'corrupt_tree': 'corrupt_tree:v0.0.1',
        'fastq_screen': 'fastq_screen:v0.0.2',
        'svtyper': 'svtyper:v0.0.1'
    }

    return {'docker': docker_images}


def get_reference_data(reference, rootdir):
    # salmon
    if reference == 'GCF_002021735':
        ref_genome = os.path.join(rootdir, 'salmon/GCF_002021735.1_Okis_V1_genomic.fna')
    elif reference == "grch37":
        classifier_training_data = os.path.join(rootdir, 'human/classifier_training_data.h5')
        gc_wig_file = {
            500000: os.path.join(rootdir, 'human/GRCh37-lite.gc.ws_500000.wig'),
            1000: os.path.join(rootdir, 'human/GRCh37-lite.gc.ws_1000.wig'),
        }
        map_wig_file = {
            500000: os.path.join(rootdir, 'human/GRCh37-lite.map.ws_125_to_500000.wig'),
            1000: os.path.join(rootdir, 'human/GRCh37-lite.map.ws_1000.wig'),
        }
        exclude_list = os.path.join(rootdir, 'human/repeats.satellite.regions')
        ref_genome = os.path.join(rootdir, 'human/GRCh37-lite.fa')
        gc_windows = os.path.join(rootdir, 'human/gc_windows.txt')
        copynumber_ref_data = os.path.join(rootdir, 'human/')
        chrom_info_filename = os.path.join(rootdir, 'human/chromInfo.txt.gz')
        chromosomes = get_chromosomes('grch37')
        destruct_ref_data = os.path.join(rootdir, 'human/')
        destruct_gtf_file = os.path.join(rootdir, 'human/GRCh37-lite.gtf')
        reference_gc_qc = os.path.join(rootdir, 'human/reference_gc_grch37.csv')
        databases = {
            'mappability': {
                'local_path': os.path.join(rootdir, 'human/wgEncodeCrgMapabilityAlign50mer.bigWig'),
            },
            'cosmic': {
                'local_path': os.path.join(rootdir, 'human/cosmic_v75.vcf.gz'),
            },
            'dbsnp': {
                'local_path': os.path.join(rootdir, 'human/dbsnp_b146_GRCh37p13.vcf.gz'),
            },
            'snpeff': {
                'local_path': os.path.join(rootdir, 'snpeff/data/')
            }
        }

    else:
        classifier_training_data = os.path.join(rootdir, 'mouse/classifier_training_data.h5')
        gc_wig_file = {
            500000: os.path.join(rootdir, 'mouse/mm10_build38_mouse.gc.ws_500000.wig'),
            1000: '/data/not/available',
        }
        map_wig_file = {
            500000: os.path.join(rootdir, 'mouse/mm10_build38_mouse.map.rl_125_ws_500000.wig'),
            1000: '/data/not/available',
        }
        exclude_list = None
        ref_genome = os.path.join(rootdir, 'mouse/mm10_build38_mouse.fasta')
        gc_windows = None
        copynumber_ref_data = os.path.join(rootdir, 'mouse/')
        chrom_info_filename = '/data/not/available'
        chromosomes = get_chromosomes('mm10')
        destruct_ref_data = os.path.join(rootdir, 'mouse/')
        destruct_gtf_file = os.path.join(rootdir, 'mouse/mm10_build38_mouse.gtf')
        reference_gc_qc = None
        databases = {
            'mappability': {
                'local_path': None,
            },
            'cosmic': {
                'local_path': None,
            },
            'dbsnp': {
                'local_path': None,
            },
            'snpeff': {
                'local_path': None
            }
        }

    return locals()


def get_reference_dir(cluster):
    if cluster == "shahlab":
        reference_dir = '/shahlab/pipelines/reference/singlecellpipeline'
    elif cluster == "juno":
        reference_dir = '/juno/work/shah/reference/singlecellpipeline'
    else:
        reference_dir = '/refdata'

    return reference_dir


def get_cluster_reference_data(refdir, reference):
    return get_reference_data(reference, refdir)


def get_chromosomes(reference):
    if reference == 'grch37':
        return [str(val) for val in range(1, 23)] + ['X', 'Y']
    elif reference == 'mm10':
        return [str(val) for val in range(1, 20)] + ['X', 'Y']
    else:
        raise Exception("unknown reference genome type {}".format(reference))


single_cell/config/pipeline_config.py
'''
Created on Jun 6, 2018

@author: dgrewal
'''
import collections
import copy

import yaml
from single_cell.config import config_reference


def override_config(config, override):
    def update(d, u):
        for k, v in u.items():
            if isinstance(v, collections.Mapping):
                d[k] = update(d.get(k, {}), v)
            else:
                d[k] = v
        return d

    if not override:
        return config

    cfg = update(config, override)

    return cfg


def get_config_params(override=None):
    input_params = {
        "cluster": "azure", "aligner": "bwa-mem", "refdir": None,
        "reference": "grch37", "smoothing_function": "modal",
        "bin_size": 500000, "copynumber_bin_size": 1000,
        'memory': {'high': 16, 'med': 6, 'low': 2},
        'version': None
    }

    input_params = override_config(input_params, override)

    return input_params


def write_config(params, filepath):
    with open(filepath, 'w') as outputfile:
        yaml.safe_dump(params, outputfile, default_flow_style=False)


def get_hmmcopy_params(reference_dir, reference, binsize, smoothing_function, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    docker_containers = config_reference.containers(version)['docker']
    docker_containers = {
        'single_cell_pipeline': docker_containers['single_cell_pipeline'],
        'hmmcopy': docker_containers['hmmcopy']
    }

    params = {
        'multipliers': [1, 2, 3, 4, 5, 6],
        'map_cutoff': 0.9,
        'bin_size': binsize,
        'e': 0.999999,
        'eta': 50000,
        'g': 3,
        'lambda': 20,
        'min_mqual': 20,
        'nu': 2.1,
        'num_states': 12,
        's': 1,
        'strength': 1000,
        'kappa': '100,100,700,100,25,25,25,25,25,25,25,25',
        'm': '0,1,2,3,4,5,6,7,8,9,10,11',
        'mu': '0,1,2,3,4,5,6,7,8,9,10,11',
        'smoothing_function': smoothing_function,
        'exclude_list': referencedata['exclude_list'],
        'gc_wig_file': referencedata['gc_wig_file'][binsize],
        'map_wig_file': referencedata['map_wig_file'][binsize],
        'chromosomes': referencedata['chromosomes'],
        'ref_genome': referencedata['ref_genome'],
        'docker': docker_containers,
        'igv_segs_quality_threshold': 0.75,
        'memory': {'med': 6},
        'good_cells': [
            ['median_hmmcopy_reads_per_bin', 'ge', 50],
            ['is_contaminated', 'in', ['False', 'false', False]],
        ]
    }

    return {"hmmcopy": params}


def get_align_params(reference_dir, reference, aligner, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)
    refdata_callback = config_reference.get_cluster_reference_data

    docker_containers = config_reference.containers(version)['docker']
    docker_containers = {
        'single_cell_pipeline': docker_containers['single_cell_pipeline'],
        'fastqc': docker_containers['fastqc'],
        'samtools': docker_containers['samtools'],
        'bwa': docker_containers['bwa'],
        'picard': docker_containers['picard'],
        'trimgalore': docker_containers['trimgalore'],
        'fastq_screen': docker_containers['fastq_screen'],
    }

    params = {
        'ref_genome': referencedata['ref_genome'],
        'docker': docker_containers,
        'memory': {'med': 6},
        'aligner': aligner,
        'adapter': 'CTGTCTCTTATACACATCTCCGAGCCCACGAGAC',
        'adapter2': 'CTGTCTCTTATACACATCTGACGCTGCCGACGA',
        'picard_wgs_params': {
            "min_bqual": 20,
            "min_mqual": 20,
            "count_unpaired": False,
        },
        'chromosomes': referencedata['chromosomes'],
        'gc_windows': referencedata['gc_windows'],
        'ref_type': reference,
        'fastq_screen_params': {
            'strict_validation': True,
            'filter_contaminated_reads': False,
            'aligner': 'bwa',
            'genomes': [
                {'name': 'grch37', 'path': refdata_callback(reference_dir, 'grch37')['ref_genome']},
                {'name': 'mm10', 'path': refdata_callback(reference_dir, 'mm10')['ref_genome']},
                {'name': 'salmon', 'path': refdata_callback(reference_dir, 'GCF_002021735')['ref_genome']},
            ]
        }
    }

    return {"alignment": params}


def get_annotation_params(reference_dir, reference, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    docker_containers = config_reference.containers(version)['docker']
    docker_containers = {
        'single_cell_pipeline': docker_containers['single_cell_pipeline'],
        'cell_cycle_classifier': docker_containers['cell_cycle_classifier'],
        'corrupt_tree': docker_containers['corrupt_tree']
    }

    params = {
        'docker': docker_containers,
        'memory': {'med': 6},
        'classifier_training_data': referencedata['classifier_training_data'],
        'reference_gc': referencedata['reference_gc_qc'],
        'chromosomes': referencedata['chromosomes'],
        'num_states': 12,
        'map_cutoff': 0.9,
        'corrupt_tree_params': {
            'neighborhood_size': 2,
            'lower_fraction': 0.05,
            'engine_nchains': 1,
            'engine_nscans': 10000,
            'model_fpr_bound': 0.1,
            'model_fnr_bound': 0.5
        },
        'good_cells': [
            ['quality', 'ge', 0.75],
            ['experimental_condition', 'notin', ["NTC", "NCC", "gDNA", "GM"]],
            ['cell_call', 'in', ['C1']],
            ['is_contaminated', 'in', ['False', 'false', False]],
        ]
    }

    return {"annotation": params}


def get_aneufinder_params(reference_dir, reference, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    docker_containers = config_reference.containers(version)['docker']
    params = {
        'memory': {'med': 6},
        'docker': {
            'single_cell_pipeline': docker_containers['single_cell_pipeline'],
            'aneufinder': docker_containers['aneufinder'],
        },
        'chromosomes': referencedata['chromosomes'],
        'ref_genome': referencedata['ref_genome']
    }

    return {'aneufinder': params}


def get_merge_bams_params(reference_dir, reference, cluster, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    if cluster in ['azure', 'aws']:
        one_split_job = True
    else:
        one_split_job = False

    docker_containers = config_reference.containers(version)['docker']
    params = {
        'memory': {'low': 4, 'med': 6, 'high': 16},
        'max_cores': 8,
        'docker': {
            'single_cell_pipeline': docker_containers['single_cell_pipeline'],
            'samtools': docker_containers['samtools']
        },
        'ref_genome': referencedata['ref_genome'],
        'split_size': 10000000,
        'chromosomes': referencedata['chromosomes'],
        'one_split_job': one_split_job
    }
    return {'merge_bams': params}


def get_split_bam_params(reference_dir, reference, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    docker_containers = config_reference.containers(version)['docker']

    params = {
        'memory': {'low': 4, 'med': 6, 'high': 16},
        'max_cores': 8,
        'docker': {
            'single_cell_pipeline': docker_containers['single_cell_pipeline'],
            'samtools': docker_containers['samtools']
        },
        'ref_genome': referencedata['ref_genome'],
        'split_size': 10000000,
        'chromosomes': referencedata['chromosomes'],
        'one_split_job': True
    }

    return {'split_bam': params}


def get_germline_calling_params(reference_dir, reference, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    docker_containers = config_reference.containers(version)['docker']

    params = {
        'memory': {'low': 4, 'med': 6, 'high': 16},
        'max_cores': 8,
        'docker': {
            'single_cell_pipeline': docker_containers['single_cell_pipeline'],
            'samtools': docker_containers['samtools'],
            'vcftools': docker_containers['vcftools'],
            'snpeff': docker_containers['snpeff'],
        },
        'ref_genome': referencedata['ref_genome'],
        'chromosomes': referencedata['chromosomes'],
        'split_size': 10000000,
        'databases': {
            'mappability': {
                'url': 'http://hgdownload-test.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/release3'
                       '/wgEncodeCrgMapabilityAlign50mer.bigWig',
                'local_path': referencedata['databases']['mappability']['local_path'],
            },
            'snpeff': {
                "db": 'GRCh37.75',
                "data_dir": referencedata['databases']['snpeff']['local_path']
            },
        },
    }

    return {'germline_calling': params}


def get_variant_calling_params(reference_dir, reference, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    docker_containers = config_reference.containers(version)['docker']

    status_data = {
        'kwargs': {
            'split_size': 10000000
        }
    }

    params = {
        'memory': {'low': 4, 'med': 6, 'high': 16},
        'max_cores': 8,
        'docker': {
            'single_cell_pipeline': docker_containers['single_cell_pipeline'],
            'vcftools': docker_containers['vcftools'],
            'strelka': docker_containers['strelka'],
            'mutationseq': docker_containers['mutationseq'],
        },
        'ref_genome': referencedata['ref_genome'],
        'chromosomes': referencedata['chromosomes'],
        'use_depth_thresholds': True,
        'split_size': 10000000,
        'cosmic_status': copy.deepcopy(status_data),
        'dbsnp_status': copy.deepcopy(status_data),
        'mappability': copy.deepcopy(status_data),
        'snpeff': copy.deepcopy(status_data),
        'tri_nucleotide_context': copy.deepcopy(status_data),
        'databases': {
            'cosmic': {
                'download_method': 'sftp',
                'user_name': 'awm3@sfu.ca',
                'password': 'shahlabith',
                'host': 'sftp-cancer.sanger.ac.uk',
                'remote_paths': {
                    'coding': '/files/grch37/cosmic/v75/VCF/CosmicCodingMuts.vcf.gz',
                    'non_coding': '/files/grch37/cosmic/v75/VCF/CosmicNonCodingVariants.vcf.gz',
                },
                'local_path': referencedata['databases']['cosmic']['local_path'],
            },
            'dbsnp': {
                'url': 'ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b146_GRCh37p13/VCF/common_all_20151104.vcf.gz',
                'local_path': referencedata['databases']['dbsnp']['local_path'],
            },
            'mappability': {
                'url': 'http://hgdownload-test.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/release3'
                       '/wgEncodeCrgMapabilityAlign50mer.bigWig',
                'local_path': referencedata['databases']['mappability']['local_path'],
            },
            'ref_genome': {
                'url': 'http://www.bcgsc.ca/downloads/genomes/9606/hg19/1000genomes/bwa_ind/genome/GRCh37-lite.fa',
                'local_path': referencedata['ref_genome'],
            },
            'snpeff': {
                "db": 'GRCh37.75',
                "data_dir": referencedata['databases']['snpeff']['local_path']
            },
        },
        'museq_params': {
            'threshold': 0.5,
            'verbose': True,
            'purity': 70,
            'coverage': 4,
            'buffer_size': '2G',
            'mapq_threshold': 10,
            'indl_threshold': 0.05,
            'normal_variant': 25,
            'tumour_variant': 2,
            'baseq_threshold': 10,
        }

    }

    return {'variant_calling': params}


def get_copy_number_calling_params(reference_dir, reference, binsize, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    docker_containers = config_reference.containers(version)['docker']

    params = {
        'memory': {'low': 4, 'med': 6, 'high': 16},
        'ref_genome': referencedata['ref_genome'],
        'chromosomes': referencedata['chromosomes'],
        'split_size': 10000000,
        'max_cores': None,
        'chromosomes': referencedata['chromosomes'],
        'extract_seqdata': {},
        'ref_data_dir': referencedata['copynumber_ref_data'],
        'docker': {
            'single_cell_pipeline': docker_containers['remixt'],
            'titan': docker_containers['titan']
        },
        'titan_params': {
            "normal_contamination": [0.2, 0.4, 0.6, 0.8],
            'num_clusters': [1, 2],
            'ploidy': [1, 2, 3, 4],
            'chrom_info_filename': referencedata['chrom_info_filename'],
            'window_size': binsize,
            'gc_wig': referencedata['gc_wig_file'][binsize],
            'mappability_wig': referencedata['gc_wig_file'][binsize],
        }
    }

    return {'copy_number_calling': params}


def get_infer_haps_params(reference_dir, reference, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    docker_containers = config_reference.containers(version)['docker']

    params = {
        'memory': {'low': 4, 'med': 6, 'high': 16},
        'max_cores': None,
        'chromosomes': referencedata['chromosomes'],
        'extract_seqdata': {
            'genome_fasta_template': referencedata['ref_genome'],
            'genome_fai_template': referencedata['ref_genome'] + '.fai',
        },
        'ref_data_dir': referencedata['copynumber_ref_data'],
        'docker': {
            'single_cell_pipeline': docker_containers['single_cell_pipeline'],
            'remixt': docker_containers['remixt'],
        },
    }

    return {'infer_haps': params}


def get_count_haps_params(reference_dir, reference, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    docker_containers = config_reference.containers(version)['docker']

    params = {
        'memory': {'low': 4, 'med': 6, 'high': 16},
        'max_cores': None,
        'chromosomes': referencedata['chromosomes'],
        'extract_seqdata': {
            'genome_fasta_template': referencedata['ref_genome'],
            'genome_fai_template': referencedata['ref_genome'] + '.fai',
        },
        'ref_data_dir': referencedata['copynumber_ref_data'],
        'docker': {
            'single_cell_pipeline': docker_containers['single_cell_pipeline'],
            'remixt': docker_containers['remixt'],
        },
    }

    return {'count_haps': params}


def get_breakpoint_params(reference_dir, reference, version):
    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    docker_containers = config_reference.containers(version)['docker']

    params = {
        'memory': {'low': 4, 'med': 6, 'high': 16},
        'ref_data_directory': referencedata['destruct_ref_data'],
        'destruct_config': {
            'genome_fasta': referencedata['ref_genome'],
            'genome_fai': referencedata['ref_genome'] + '.fai',
            'gtf_filename': referencedata['destruct_gtf_file'],
        },
        'docker': {
            'single_cell_pipeline': docker_containers['single_cell_pipeline'],
            'destruct': docker_containers['destruct'],
            'lumpy': docker_containers['lumpy'],
            'samtools': docker_containers['samtools'],
        },
    }

    return {'breakpoint_calling': params}


def get_sv_genotyping_params(reference_dir, reference, version):
    docker_containers = config_reference.containers(version)['docker']

    referencedata = config_reference.get_cluster_reference_data(reference_dir, reference)

    params = {
        'ref_genome': referencedata['ref_genome'],
        'memory': {'low': 4, 'med': 6, 'high': 16},
        'docker': {
            'single_cell_pipeline': docker_containers['single_cell_pipeline'],
            'svtyper': docker_containers['svtyper']
        },
    }
    return {'sv_genotyping': params}


def get_singlecell_pipeline_config(config_params, override=None):
    reference = config_params["reference"]
    reference_dir = config_params['refdir']
    version = config_params['version']
    cluster = config_params["cluster"]
    if not reference_dir:
        reference_dir = config_reference.get_reference_dir(cluster)

    params = {}

    params.update(
        get_hmmcopy_params(
            reference_dir, reference, config_params["bin_size"],
            config_params["smoothing_function"], version
        )
    )

    params.update(
        get_align_params(
            reference_dir, reference,
            config_params['aligner'],
            version
        )
    )

    params.update(get_annotation_params(reference_dir, reference, version))

    params.update(get_aneufinder_params(reference_dir, reference, version))

    params.update(get_merge_bams_params(reference_dir, reference, cluster, version))

    params.update(get_split_bam_params(reference_dir, reference, version))

    params.update(get_germline_calling_params(reference_dir, reference, version))

    params.update(get_variant_calling_params(reference_dir, reference, version))

    params.update(get_copy_number_calling_params(
        reference_dir, reference, config_params['copynumber_bin_size'], version)
    )

    params.update(get_infer_haps_params(reference_dir, reference, version))

    params.update(get_count_haps_params(reference_dir, reference, version))

    params.update(get_breakpoint_params(reference_dir, reference, version))

    params.update(get_sv_genotyping_params(reference_dir, reference, version))

    params = override_config(params, override)

    return params
